Bindings rework

knah opened the pull request:
This PR includes the following changes to the input system:
* Store custom binding information per-profile and per-hand explicitly
  * Most (all?) bindings in SteamVR are specifying only one hand as input source anyway
  * This incidentally allows detecting action origin for custom bindings, which is used when possible
  * This also prevents cross-contamination of custom bindings between different profiles/controller types.
* Custom binding information and state is stored separately from actions
* Adjusted index controller grip thresholds to be in-line with SteamVR
  * This changes how the "grab" action is read as a `f32` - it's now binary 0/1 and never returns any inbetween values. A common `f32` binding type for grip seems to be "trigger" or "force_sensor", but if you have examples of games relying on "grab" binding type having a `f32` value - please let me know!
* Implemented custom thresholds for "button" bindings
  * This includes support for alternate click source, such as force; default is "value" or "click", in that order
  * This seems to be consistent with binding to clicky triggers - a "button" binding doesn't actually require a click, instead a "trigger" binding type should be used if a click is desired
  * Also implemented "touch" input for "button" bindings
* Implemented force threshold and haptic feedback for "dpad" bindings on index controller touchpad
* Simple controller profile no longer pretends to be a vive wand. I don't expect any game to actually ship with bindings for this profile, although SteamVR binding UI has it in its long controller list.
* Minor fixes for various path translation and binding parsing issues

Games tested by me so far with index controllers:
* VRC - most actions work fine, but gestures fail due to missing skeletal summary data
* H3VR - played without issues, seems to be consistent with OC/SteamVR

More testing with a more varied set of games would be welcome

Supreeeme commented:
This is a lot of changes, some of which are unrelated, and it would be easier for me to review if it was split into a few separate PRs.

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -1015,29 +1083,43 @@ fn handle_dpad_binding(
+        let mut data = extra_actions.remove(action_name).unwrap_or_default();
+
+        if data.dpad_actions.is_none() {
+            // FIXME: this steals a random action from app's action set for haptic feedback; there might be none
+            let haptic = if matches!(controller_type, ControllerType::Knuckles) && parent_path.ends_with("trackpad") {
+                let actions = actions.borrow();
+                actions.iter().filter_map(|(_, a)| if let ActionData::Haptic(h) = a { Some(h) } else { None }).find(|_| true).map(|x| x.clone())
+            } else {
+                None
+            };
```
I don't like this - just create a haptic action for it.

knah replied:
My concern here is action set priorities. The spec seemingly says nothing about how action set priorities interact with output actions, but it would be a shame to break game's haptic output in case a dpad haptic from a higher-priority action set does in fact override it.

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -1015,29 +1083,43 @@ fn handle_dpad_binding(
-            data.dpad_data = Some(DpadData {
-                parent: parent_action.clone(),
+            data.dpad_actions = Some(DpadActions {
+                direction: parent_action.clone(),
```
Why is this named `direction`?

knah replied:
A "dpad" binding has two input actions - the direction and the (possibly optional) click. This naming seems to be cleaner, especially in the context of `DpadActions`

Supreeeme replied:
A direction to me implies an actual direction (north, south, east, west). A "direction" action is unclear, especially considering this is a Vector2f action which doesn't return a specific direction but rather 2 floats.

knah replied:
I'm up for naming it something more descriptive than either "direction" or "parent_action", perhaps something straightforward like "xy"?

Supreeeme replied:
I can live with that.

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -1148,16 +1238,32 @@ fn translate_warn(action: &str) -> impl FnOnce(&InvalidActionPath) + '_ {
+    };
+
+    let path = instance.string_to_path(hand_prefix).ok();
+    path.map(|x| if x == xr::Path::NULL { None } else { Some(x) }).flatten()
```
```suggestion
    path.and_then(|x| if x == xr::Path::NULL { None } else { Some(x) })
```

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -1204,54 +1310,131 @@ fn handle_sources(
+                    let translated = if let Ok(translated) = path_translator(&format!("{path}/{target}"))
+                        .inspect_err(|e| info!("Falling back to click for {output} ({})", e.0)) {
+                        translated
+                    } else {
+                        let Ok(translated) = path_translator(&format!("{path}/click"))
+                            .inspect_err(translate_warn(output)) else {
+                            continue;
+                        };
+                        translated
+                    };
```
```suggestion
                    } else if let Ok(translated) = path_translator(&format!("{path}/click"))
                            .inspect_err(translate_warn(output))  {
                            translated
                        };
                    } else {
                        continue;
                    };
```

Supreeeme commented on src/input/custom_bindings.rs:
```diff
@@ -558,64 +472,54 @@ mod tests {
     }
 
     #[test]
-    fn grab_per_hand() {
```
It seems this test has been completely removed?

knah replied:
Oops, it got eaten by the other test, I put it back in.

Supreeeme commented on src/input/profiles.rs:
```diff
@@ -23,6 +23,7 @@ pub trait InteractionProfile: Sync + Send {
     fn offset_grip_pose(&self, pose: xr::Posef) -> xr::Posef {
         pose
     }
+    fn controller_type(&self) -> ControllerType;
```
Why add this? This association is already handled by `Profiles::get`.

knah replied:
I reverted the commit that includes this and source the `ControllerType` from caller instead.

knah commented on src/input.rs:
```diff
@@ -185,6 +245,18 @@ macro_rules! get_action_from_handle {
+            Ok(action) => action,
+            Err(e) => return e,
+        };
+    };
```
The reason for separate branch is exposing `loaded` to the macro user. I replaced the original branch with a use of this one.

knah commented on src/input.rs:
```diff
@@ -125,6 +125,60 @@ impl<C: openxr_data::Compositor> Input<C> {
+            };
+
+            if state.is_active && (!best_state.map(|x| x.is_active).unwrap_or(false) || state.current_state && !best_state.map(|x| x.current_state).unwrap_or(false)) {
+                best_state = Some(state);
```
This could have been a leftover from my initial draft where custom bindings were `f32`. Added a `break`.

knah commented on src/input/action_manifest.rs:
```diff
@@ -25,6 +25,21 @@ use std::{
 };
 
 impl<C: openxr_data::Compositor> Input<C> {
+    fn action_map_to_secondary<T>(&self, map: HashMap<String, T>) -> SecondaryMap<ActionKey, T> {
+        let mut act_guard = self.action_map.write().unwrap();
+
+        map.into_iter()
+            .map(|(name, action)| {
+                let key = act_guard
+                    .iter()
+                    .find_map(|(key, super::Action { path })| (*path == name).then_some(key))
+                    .unwrap_or_else(|| act_guard.insert(super::Action { path: name }));
+
+                (key, action)
+            })
+            .collect()
+    }
```
I'm not quite sure which one - `action_map_to_secondary` uses the lock-protected map from `self`, and the lambda in `map` uses captured variables too.

knah commented:
> This is a lot of changes, some of which are unrelated, and it would be easier for me to review if it was split into a few separate PRs.

I hope that the separation of changes into individual well-defined commits does make reviewing easier. GH PR UI allows viewing single-commit diffs via a selector in the top-left.


I fixed up a few more occurrences of `map().flatten()` in addition to your suggestion.

Let me know if you want me to `rebase --autosquash` to clean up commit history before merge

Supreeeme reviewed (approved):
Looks ok so far, just minor things remaining. Tested it in a few of my games and they seemed fine. Hopefully the test coverage is enough. Be sure to run `cargo fmt` and `cargo clippy`.

knah commented on src/input/action_manifest.rs:
```diff
@@ -190,12 +222,14 @@ struct DefaultBindings {
     Knuckles,
     OculusTouch,
+    // AFAIK there's no bindings for it, but also there's no reason to suggest bindings for it
+    SimpleController,
```
I see what happened there - the original approach had maps keyed with `ControllerType` instead of `xr::Path`, so I wanted to have profiles fully separate. I restored the existing logic - although it does spam the log quite a bit with mismatched bindings.

knah commented:
I squashed the fixups and did both `fmt` and `clippy`. There's leftover warnings about functions having too many arguments that I don't quite believe to be useful - shoving that same data into a `struct` won't help readability I'd say.

There was also a lost dpad haptic due to missing subaction paths on dpad bindings - I fixed that one too.

Supreeeme commented:
> There's leftover warnings about functions having too many arguments that I don't quite believe to be useful
Add `#[allow(clippy::too_many_arguments)]` to these.
If you don't mind fixing the merge conflict I will merge this - think it's just a test that got removed.

knah commented:
I did end up splitting some state/code into separate file to hopefully make it easier to read. I also rebased it on latest main.

knah commented:
Hopefully last changes I'll shove in here, these fix issues seen in Alyx, bringing it to playable on index and touch.

Supreeeme commented on src/input/action_manifest_helpers.rs:
```diff
@@ -0,0 +1,604 @@
+use crate::input::action_manifest::{
```
I'd prefer if this file were a module under `action_manifest`, structured like this:
```
action_manifest/helpers.rs
action_manifest.rs
```

Supreeeme commented on src/input/action_manifest_helpers.rs:
```diff
@@ -0,0 +1,604 @@
+        float_name_with_as
+    }
+
+    pub fn get_or_create_grip_action_pair(
```
```suggestion
    pub fn get_or_create_grab_action_pair(
```
The type is "grab" in the binding files.

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -671,11 +694,11 @@ struct DpadInput {
-    deadzone_pct: FromString<u8>,
-    overlap_pct: FromString<u8>,
-    sticky: FromString<bool>,
+pub struct DpadParameters {
+    pub sub_mode: DpadSubMode,
+    pub deadzone_pct: FromString<u8>,
+    pub overlap_pct: FromString<u8>,
+    pub sticky: FromString<bool>,
```
If `action_manifest_helpers` is made a submodule then this stuff doesn't have to be made pub.

Supreeeme commented on src/input/action_manifest_helpers.rs:
```diff
@@ -0,0 +1,604 @@
+        // Remove lifetime
+        let click_or_touch = activator_action.cloned();
+
+        let haptic_data = if use_force {
+            // the need for haptic coincides with force-using dpads for now
+            let hand_path = get_hand_prefix(parent_path)
+                .and_then(|x| string_to_path(&format!("{x}/output/haptic")));
+            let haptic_key = format!("{parent_path}-{action_set_name}-haptic");
+            hand_path.map(|hand_path| {
+                let action = self.actions.entry(haptic_key.clone()).or_insert_with(|| {
+                    let haptic_name = format!("xrizer-dpad-haptic{len}");
+                    let localized = format!("XRizer dpad haptic ({len})");
+
+                    ActionData::Haptic(
+                        action_set
+                            .create_action(&haptic_name, &localized, &self.hands)
+                            .unwrap(),
+                    )
+                });
+
+                let ActionData::Haptic(action) = action else {
+                    unreachable!();
+                };
+                DpadHapticData {
+                    action: action.clone(),
+                    key: haptic_key,
+                    binding: hand_path,
+                }
+            })
+        } else {
+            None
+        };
```
```rust
use_force.then(|| {
// ...
});
```

Supreeeme commented on src/input/profiles.rs:
```diff
@@ -98,6 +98,7 @@ impl Profiles {
                 (ControllerType::Knuckles, &Knuckles),
                 (ControllerType::OculusTouch, &Touch),
                 (ControllerType::ViveController, &SimpleController),
```
Did you mean to delete this line?

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -4,7 +4,7 @@ use super::{
     ActionData, ActionKey, BoundPoseType, Input,
 };
-use crate::input::action_manifest_helpers::{
+use crate::input::action_manifest::helpers::{
```
```suggestion
use helpers::{
```

Supreeeme commented on src/input/action_manifest.rs:
```diff
@@ -203,12 +235,13 @@ struct DefaultBindings {
     ViveController,
     Knuckles,
     OculusTouch,
+    KhrSimpleController,
```
I still don't like this - I don't understand why this is being added when no real game is going to do this.

knah replied:
Alright, I dropped it completely, together with a commit related to it. Also rebased on latest main.

Supreeeme commented:
Okay, I think I'm satisfied with how this looks now. Thanks!
